onmt/transforms/misc.py

```diff
@@ -41,6 +43,84 @@ def _repr_args(self):
         )
 
 
+@register_transform(name="joiner_dropout")
+class JoinerDropoutTransform(Transform):
+    """Disjoin joiner with probability dropout"""
+
+    def __init__(self, opts):
+        super().__init__(opts)
+
+    @classmethod
+    def add_options(cls, parser):
+        """Avalilable options relate to this Transform."""
+        group = parser.add_argument_group("Transform/JoinerDropout")
+        group.add(
+            "--src_joiner_dropout",
+            "-src_joiner_dropout",
+            type=float,
+            default=0.0,
+            help="Source dropout probability.",
+        )
+        group.add(
+            "--tgt_joiner_dropout",
+            "-tgt_joiner_dropout",
+            type=float,
+            default=0.0,
+            help="Target dropout probability.",
+        )
+
+    def _parse_opts(self):
+        self.src_joiner_dropout = self.opts.src_joiner_dropout
+        self.tgt_joiner_dropout = self.opts.tgt_joiner_dropout
+
+    def dropout_separate_joiner(self, seq, side="src"):
+        out_seq = []
+        dropout = (
+            self.src_joiner_dropout
+            if side == "src"
+            else self.tgt_joiner_dropout
+        )
+        for elem in seq:
+            if elem == SubwordMarker.JOINER:
+                continue
+            if elem.startswith(SubwordMarker.JOINER):
+                if random.random() < dropout:
+                    out_seq.append(SubwordMarker.JOINER)
+                    elem = elem[1:]
+            if len(elem) > 1 and elem.endswith(SubwordMarker.JOINER):
+                if random.random() < dropout:
+                    out_seq.append(elem[:-1])
+                    out_seq.append(SubwordMarker.JOINER)
+                else:
+                    out_seq.append(elem)
+            else:
+                out_seq.append(elem)
+        return out_seq
+
+    def apply(self, example, is_train=False, stats=None, **kwargs):
+        """Return None if too long else return as is."""
+        if not is_train:
+            return example
+        else:
+            src_out = self.dropout_separate_joiner(example["src"], "src")
+            example["src"] = src_out
+            if self.opts.model_task == ModelTask.LANGUAGE_MODEL:
```

Comment:
Currently this won't work on build_vocab, model_task is a model parameter.
Same issue in tokenizers.py

Reply:
`is_train` should be false anyways in `build_vocab`, no?